mcs/class/System.Windows.Forms/System.Windows.Forms/Control.cs

```diff
@@ -1393,8 +1393,8 @@ private Control FindControlToInvokeOn ()
 			}
 
 			if (background_image == null) {
-				if (!tbstyle_flat) {
-					Rectangle paintRect = pevent.ClipRectangle;
+				if (!tbstyle_flat && pevent.ClipRectangle != null) {
+					Rectangle paintRect = pevent.ClipRectangle; 					
```

Comment:
Isn't ClipRectangle a structure (ie. value type)? Value types are not nullable.

Reply:
```c#
    Rectangle a = default(Rectangle);
    Console.WriteLine(a == null);
```

returns false. I remember reading somewhere that it's not an outright compilation error because some compatibility dating back all the way to .NET 1.x. Cannot find the article though.

Reply:
I don't think that's the actual problem causing the crash though. It looks more likely to be some control disposing the `Graphics` object in the `PaintEventArgs`, or a similar issue. Unfortunately that's impossible to verify based just on the stack trace. I'd hesitate to make this change unless a test application / unit test is provided that can confirm the issue is solved.

Reply:
Yes, the problem may be in some Control, but it is not ok that Windows Forms is crashing... 
My application is running 4hours and then suddenly this error occurs. 

i really suggest doing a structure like

            Rectangle r=new Rectangle(0,0,0,0);
            if (r!=null && !r.IsEmpty)
            {
               ....
            }

Reply:
If my suspicion is correct then that won't help at all. It would crash exactly the same way.

Reply:
https://dotnetfiddle.net/aZn0i6

Reply:
Isn´t it better to ignore a invalid System.Windows.Forms.Message& ? than giving this up to System.Drawing.Graphics.FillRectangle ?

I assume the main reason could be video driver related or has to do with X system

Reply:
It is not invalid. The drawing is processed by the libgdiplus and in turn by the Cairo library, not by a video driver. If it doesn't crash alone with empty rectangle there's no reason to believe it would crash when wrapped in a more complex code.

You are misinterpreting the cause of the error and you have not submitted any test code that would show the ClipRectangle is related in any way to the error.

Reply:
If I run test from #8922 with 

`mono  --trace=T:System.Drawing.Graphics Monitor.exe`

trace will gives this part of output

`[0x7f83a4236780: 1,34770 2] LEAVE: System.Drawing.Graphics:SetClip (System.Drawing.Region,System.Drawing.Drawing2D.CombineMode)
[0x7f83a4236780: 1,34771 1] LEAVE: System.Drawing.Graphics:set_Clip (System.Drawing.Region)
[0x7f83a4236780: 1,34772 0] LEAVE: (wrapper remoting-invoke-with-check) System.Drawing.Graphics:set_Clip (System.Drawing.Region)
[0x7f83a4236780: 1,34840 0] ENTER: (wrapper remoting-invoke-with-check) System.Drawing.Graphics:FillRectangle (System.Drawing.Brush,System.Drawing.Rectangle)(this:0x7f83a22a9560[System.Drawing.Graphics Monitor.exe], [System.Drawing.SolidBrush:0x7f83a22a8a80], [], )
[0x7f83a4236780: 1,34842 1] ENTER: System.Drawing.Graphics:FillRectangle (System.Drawing.Brush,System.Drawing.Rectangle)(this:0x7f83a22a9560[System.Drawing.Graphics Monitor.exe], [System.Drawing.SolidBrush:0x7f83a22a8a80], [], )
[0x7f83a4236780: 1,34845 2] ENTER: System.Drawing.Graphics:FillRectangle (System.Drawing.Brush,int,int,int,int)(this:0x7f83a22a9560[System.Drawing.Graphics Monitor.exe], [System.Drawing.SolidBrush:0x7f83a22a8a80], 0, 0, 2147483647, 29, )
[0x7f83a4236780:] EXCEPTION handling: System.Exception: Generic Error [GDI+ status: GenericError]
[0x7f83a4236780: 1,34906 3] ENTER: (wrapper remoting-invoke-with-check) System.Drawing.Graphics:Dispose ()(this:0x7f83a22a9560[System.Drawing.Graphics Monitor.exe], )
`

Invalid params for Rectange ctor is 0, 0, 2147483647, 29, gived before GenericError exception
On .net windows there will be OutOfRangeException
My system is

	mono -V
	Mono JIT compiler version 5.21.0 (master/d319f72d0dd Вс окт  7 12:53:58 MSK 2018)
	Copyright (C) 2002-2014 Novell, Inc, Xamarin Inc and Contributors. www.mono-project.com
	TLS:           __thread
	SIGSEGV:       altstack
	Notifications: epoll
	Architecture:  amd64
	Disabled:      none
	Misc:          softdebug 
	Interpreter:   yes
	LLVM:          supported, not enabled.
	Suspend:       hybrid
	GC:            sgen (concurrent by default)

	uname -a
	Linux sancheolz-VirtualBox 4.15.0-36-generic #39-Ubuntu SMP Mon Sep 24 16:19:09 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
	
	Ubuntu 18.04.1 LTS 
	gnome 3.28.2

Reply:
manipulating a control from a timer_tick without a lock of the control can cause this.

Reply:
@sancheolz The issue was not linked from the PR so I didn't see the test case. That indeed confirms my suspicion that the cause is not an empty ClipRectangle and hence this PR would not help. There seems to be some invalid value sneaking in though (int.MaxValue), which deserves some attention.

Reply:
There are so many pointing exactly to this.  These Invalid params for rectangles like  ctor is 0, 0, 2147483647, 29 should be ignored

google for "at System.Drawing.Graphics.FillRectangle"

Reply:
In mean call 
`System.Drawing.Graphics:FillRectangle (Brushes.AliceBlue ,0, 2147483647, 29);`
, not just call 
`new Rectangle(0, 2147483647, 29);`
The https://dotnetfiddle.net/HA0hxF on .net on windows will give
```
Run-time exception (line 10): Overflow error.

Stack Trace:

[System.OverflowException: Overflow error.]
   at System.Drawing.Graphics.CheckErrorStatus(Int32 status)
   at System.Drawing.Graphics.FillRectangle(Brush brush, Int32 x, Int32 y, Int32 width, Int32 height)
   at System.Drawing.Graphics.FillRectangle(Brush brush, Rectangle rect)
   at Program.Main() :line 10
```